From 59cc65411f02c7e39a270fda3ecb4966d7b48d41 Mon Sep 17 00:00:00 2001
From: Matthew Noorenberghe <github@matthew.noorenberghe.com>
Date: Wed, 23 Jan 2019 19:40:38 -0800
Subject: [PATCH] plugin.php: Fix XSS and directory traversal bugs. Fixes #2436

This view seems like dead code so maybe it should be removed instead.
---
 web/skins/classic/views/plugin.php | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/web/skins/classic/views/plugin.php b/web/skins/classic/views/plugin.php
index 7b4cf85ca6..814416058e 100644
--- a/web/skins/classic/views/plugin.php
+++ b/web/skins/classic/views/plugin.php
@@ -36,7 +36,8 @@
    return;
 }
 $monitor = dbFetchMonitor ( $mid );
-$plugin = $_REQUEST['pl'];
+// Only allow certain filename characters (not including a period) to prevent directory traversal.
+$plugin = preg_replace('/[^-a-zA-Z0-9]/', '', $_REQUEST['pl']);
 
 $plugin_path = dirname(ZM_PLUGINS_CONFIG_PATH)."/".$plugin;
 
@@ -103,7 +104,7 @@ function pLang($name)
 <body>
   <div id="page">
     <div id="header">
-      <h2><?php echo translate('Monitor') ?> <?php echo $monitor['Name'] ?> - <?php echo translate('Zone') ?> <?php echo $newZone['Name'] ?> - <?php echo translate('Plugin') ?> <?php echo $plugin ?></h2>
+      <h2><?php echo translate('Monitor') ?> <?php echo $monitor['Name'] ?> - <?php echo translate('Zone') ?> <?php echo $newZone['Name'] ?> - <?php echo translate('Plugin') ?> <?php echo validHtmlStr($plugin) ?></h2>
     </div>
     <div id="content">
       <form name="pluginForm" id="pluginForm" method="post" action="<?php echo $_SERVER['PHP_SELF'] ?>">
@@ -111,7 +112,7 @@ function pLang($name)
         <input type="hidden" name="action" value="plugin"/>
         <input type="hidden" name="mid" value="<?php echo $mid ?>"/>
         <input type="hidden" name="zid" value="<?php echo $zid ?>"/>
-        <input type="hidden" name="pl" value="<?php echo $plugin ?>"/>
+        <input type="hidden" name="pl" value="<?php echo validHtmlStr($plugin) ?>"/>
 
         <div id="settingsPanel">
           <table id="pluginSettings" cellspacing="0">
